fix(lua): revert vim.tbl_deep_extend behavior change and document it.

clason opened the pull request:
# Problem: 

`vim.tbl_deep_extend` had an undocumented feature where arrays (integer-indexed tables) were not merged but compared literally (used for merging default and user config, where one list should
overwrite the other completely). Turns out this behavior was relied on in quite a number of plugins (even though it wasn't a robust solution even for that use case, since lists of tables (e.g., plugin specs) can be array-like as well).

# Solution: 

Revert the removal of this special feature. Check for list-like (contiguous integer indices starting from 1) instead, as this is closer to the intent. Document this behavior.

Follow-up should be to add a new (better) `vim.merge` API that covers all current use cases and deprecates these legacy functions. It should also offer more -- and more explicit -- control. (See, e.g., https://github.com/neovim/neovim/pull/15124#issuecomment-2336685680.)

justinmk commented on runtime/doc/lua.txt:
```diff
@@ -2230,6 +2230,11 @@ vim.tbl_count({t})                                           *vim.tbl_count()*
 vim.tbl_deep_extend({behavior}, {...})                 *vim.tbl_deep_extend()*
     Merges recursively two or more tables.
 
+    Only empty tables or tables that are not list-like (indexed by consecutive
```
link to  `:help lua-list`  ?

justinmk commented on test/functional/lua/vim_spec.lua:
```diff
@@ -1071,15 +1071,15 @@ describe('lua stdlib', function()
-    ]]))
-
-    matches('invalid "behavior": nil', pcall_err(exec_lua, [[return vim.tbl_deep_extend()]]))
+    matches(
```
this matches() case looks like a style-only change

clason replied:
Yes, the revert dropped a style-only change in the meantime. Will re-revert.

clason committed:
fix(lua): revert vim.tbl_extend behavior change and document it

Problem: vim.tbl_deep_extend had an undocumented feature where arrays
(integer-indexed tables) were not merged but compared literally (used
for merging default and user config, where one list should overwrite the
other completely). Turns out this behavior was relied on in quite a
number of plugins (even though it wasn't a robust solution even for that
use case, since lists of tables (e.g., plugin specs) can be array-like
as well).

Solution: Revert the removal of this special feature. Check for
list-like (contiguous integer indices) instead, as this is closer to the
intent. Document this behavior.

clason commented:
That was the only use of `isarray` in our codebase -- should I go ahead and deprecate it while I'm at it?

echasnovski commented on runtime/doc/lua.txt:
```diff
@@ -2230,6 +2230,11 @@ vim.tbl_count({t})                                           *vim.tbl_count()*
     Merges recursively two or more tables.
 
+    Only empty tables or tables that are not |lua-list|s (indexed by
+    consecutive integers starting from 1) are merged. This is useful for
```
For reference, the reverted implementation (the one present in 0.10.1) does not do this on "first level" of merging. I.e.:
 - `vim.tbl_deep_extend('force', {'a', 'b'}, {'c'})` -> `{ 'c', 'b' }`
 - `vim.tbl_deep_extend('force', {sub = {'a', 'b'} }, { sub = { 'c' } })` -> `{ sub = { 'c' } }`.

This was the whole point of #23654. Not sure if this still requires a fix or should be documented as intended.

clason replied:
This is very likely intended/expected (note that tables of tables -- like plugin specs -- are also list-like). I'll try to document this.

clason replied:
(And, yes, this is an inconsistency which would be better to fix; alas, we cannot do so without too much breakage, as it turns out. Hence we need a proper deprecation cycle, with a new API that is rational.)

folke replied:
fyi, not sure if you're talking about lazy's plugin specs, but there @echasnovski's first example would yield `{'c'}`, which does indeed make more sense. So I'm also a proponent to have this fixed (either now, or in a later iteration)

clason replied:
I was talking more generally, as an excuse not to get my hands too dirty. If you can promise to keep the general pitchforks at bay, I can look at extending `can_merge` to the top-level scope. That will likely take a bit more time since #dayjob will reassert itself powerfully soon.

folke replied:
The most important thing right now is the PR in its current state.
In the context of option/config merging, the top level tables would not be list-like, so this problem does not happen there at least.

I'm willing to work on a `vim.merge` PR that includes this fix this week (after this PR gets merged).

GitMurf replied:
If either of you need / want testing or feedback from an end user perspective as this progresses, please don't hesitate to ask!

justinmk commented on runtime/doc/lua.txt:
```diff
@@ -2230,6 +2230,12 @@ vim.tbl_count({t})                                           *vim.tbl_count()*
+    Only values that are empty tables or tables that are not |lua-list|s
+    (indexed by consecutive integers starting from 1) are merged recursively.
+    This is useful for merging nested tables like default and user
+    configurations where lists should be treated as literals (i.e., are
```
```suggestion
    configurations where nested lists should be treated as literals (i.e., are
```

clason replied:
Does it make a difference (honestly asking)?

justinmk replied:
Are top-level lists merged?

Anyway, I think the "nested tables" on the previous line perhaps is the important part, so my suggestion can be ignored.

clason replied:
> Are top-level lists merged?

No, as @echasnovski points out above. Only lists that are _values_ are special-cased.

clason replied:
(Always happy about suggestions for improvement; I'd just like to understand them to learn from them.)

justinmk commented on runtime/doc/lua.txt:
```diff
@@ -2230,6 +2230,12 @@ vim.tbl_count({t})                                           *vim.tbl_count()*
 vim.tbl_deep_extend({behavior}, {...})                 *vim.tbl_deep_extend()*
     Merges recursively two or more tables.
 
+    Only values that are empty tables or tables that are not |lua-list|s
+    (indexed by consecutive integers starting from 1) are merged recursively.
```
if we are talking about "merging recursively", maybe it's not worth mentioning empty tables?

```suggestion
    Only tables that are not |lua-list|s
    (indexed by consecutive integers starting from 1) are merged recursively.
```

clason replied:
No, these are special cased (_and_ list-like), so this is important to call out.

echasnovski replied:
The current "values that are ..." is a way to document [this behavior](https://github.com/neovim/neovim/pull/30309#discussion_r1749287011).

justinmk commented:
> That was the only use of `isarray` in our codebase -- should I go ahead and deprecate it while I'm at it?

Needs a separate PR assuming we want to actually tackle that. IMO it seems useful to keep it so that we have a programmatic definition of the "spec" for an array.

clason commented:
> Needs a separate PR assuming we want to actually tackle that.

Of course (or at least a separate, non-squashed, commit).

> IMO it seems useful to keep it so that we have a programmatic definition of the "spec" for an array. (Not to mention, if anyone finds a use-case, it would feel pretty bad that we removed it.)

Well, the idea was that adding a deprecation would be the fastest way of finding out (the hard way) ;) 

But no strong feelings either way myself.

folke reviewed (approved):
Looks good to me!

And thank you for the quick turn-around. I'm sure this will avoid lots of mysterious issues popping up all over Github :)
